[Console] Converted all ```/lib/autocomplete/**/*.js``` files to typescript

curq opened the pull request:
### Description

Converts ```lib/autocomplete``` to typescript. Part of the https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3825

### Issues Resolved

Solves https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3826

### Testing the changes

No type errors present in affected files, except from imports of other non-ts files, will be fixed by other PRs from the same group of issues (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3827). 

### Check List

- [X] All tests pass
  - [X] `yarn test:jest`
  - [X] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [X] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

curq commented:
This PR should ideally be merged after https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008 as it depends on types from it.

ananzh reviewed (changes requested):
nice job! thank you for the help~

curq commented on CHANGELOG.md:
```diff
@@ -209,6 +209,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Console] Remove unused ul element and its custom styling ([#3993](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3993))
 - Fix EUI/OUI type errors ([#3798](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3798))
 - Remove unused Sass in `tile_map` plugin ([#4110](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4110))
+- [Console] Convert /lib/autocomplete/ to TypeScript ([#4148](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148))
```
Yeah, I think it will be better description, I updated changelog

curq commented on src/plugins/console/public/lib/autocomplete/url_params.ts:
```diff
@@ -29,27 +29,12 @@
  */
 
 import _ from 'lodash';
```
I have removed lodash usage in this file, and I will remove the rest in a separate PR dedicated to refactoring console plugin.

ananzh replied:
Thank you @curq. You are definitely our star contributor!

curq commented on src/plugins/console/public/lib/autocomplete/types.ts:
```diff
@@ -0,0 +1,102 @@
+/*
```
Done.

ananzh replied:
I still see the old license. @curq did you forget to submit your commit?

curq commented on src/plugins/console/public/lib/autocomplete/types.ts:
```diff
@@ -0,0 +1,102 @@
+  autoCompleteType: string;
+  // autoCompleteSet: string[] | Term[] | null;
+  autoCompleteSet: Term[] | null;
+  createdWithToken?: Token;
```
Yes, you're right, I will change that

curq commented on src/plugins/console/public/lib/autocomplete/body_completer.ts:
```diff
@@ -308,7 +345,11 @@ export function globalsOnlyAutocompleteComponents() {
  * }
  */
-export function compileBodyDescription(endpointId, description, parametrizedComponentFactories) {
+export function compileBodyDescription(
```
I have updated the comments with cleaner version.

curq commented on src/plugins/console/public/lib/autocomplete/body_completer.ts:
```diff
@@ -308,7 +345,11 @@ export function globalsOnlyAutocompleteComponents() {
+export function compileBodyDescription(
+  endpointId: string,
+  description: Description,
+  parametrizedComponentFactories?: ParametrizedComponentFactories
```
It rechecked and it is in fact not optional. Thanks for the catch!  I updated it to required.

curq commented on src/plugins/console/public/lib/autocomplete/engine.ts:
```diff
@@ -29,8 +29,15 @@
+
+export function wrapComponentWithDefaults(
+  component: SharedComponent,
+  defaults: Record<string, unknown>
```
Yes, I looked through code and found appropriate type. Will update.

curq commented on src/plugins/console/public/lib/autocomplete/components/constant_component.ts:
```diff
@@ -30,27 +30,31 @@
-  constructor(name, parent, options) {
+  options: Term[];
+
+  constructor(name: string, parent?: SharedComponent | null, options?: string | Term[]) {
```
I updated addOption to Term | Term[]. In other two cases:
1. options as property of ConstantComponent is always array of Terms
2. in constructor it is also the same, but we have a case where the options is string, in which case it is converted to the array of terms.

curq commented on src/plugins/console/public/lib/autocomplete/components/param_component.ts:
```diff
@@ -0,0 +1,51 @@
+import { ConstantComponent } from './constant_component';
+
+export class ParamComponent extends ConstantComponent {
+  description: string | unknown;
```
Yes, after researching the ```description``` variable I've found out that it is the same as in https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2322c53d10664d68c7d2ee21a86a4e29d1919d9c/src/plugins/console/server/services/spec_definitions_service.ts#L43-L53
Here we can see that for description ```any``` is used, also here is the examples of description variable 
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2322c53d10664d68c7d2ee21a86a4e29d1919d9c/src/plugins/console/public/lib/autocomplete/__jest__/url_params.test.js#L95-L98
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2322c53d10664d68c7d2ee21a86a4e29d1919d9c/src/plugins/console/public/application/models/sense_editor/__tests__/integration.test.js#L1013-L1016
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2322c53d10664d68c7d2ee21a86a4e29d1919d9c/src/plugins/console/server/lib/spec_definitions/json/generated/cat.allocation.json#L3-L23
I have seen other examples of big complex ```description```s, but I'm not sure that they are the same as this one.

Overall I concluded that it can be a big complex object and we can't be 100% sure how it will look like.

curq commented on src/plugins/console/public/lib/autocomplete/components/shared_component.ts:
```diff
@@ -41,14 +44,14 @@ export class SharedComponent extends AutocompleteComponent {
   /* return the first component with a given name */
-  getComponent(name) {
+  getComponent(name: string): SharedComponent | undefined {
     return (this._nextDict[name] || [undefined])[0];
```
Yes, it is initialized in constructor when the SharedComponent is created, so I assumed that it is always present.

curq commented on src/plugins/console/public/lib/autocomplete/components/shared_component.ts:
```diff
@@ -41,14 +44,14 @@ export class SharedComponent extends AutocompleteComponent {
     current.push(component);
     this._nextDict[component.name] = current;
-    this.next = [].concat.apply([], _.values(this._nextDict));
+    this.next = ([] as SharedComponent[]).concat.apply([], _.values(this._nextDict));
```
Yes, I have update code

curq commented on src/plugins/console/public/lib/autocomplete/__jest__/url_params.test.ts:
```diff
@@ -63,32 +72,34 @@ describe('Url params', () => {
+    let r: string | Term = name;
     if (meta) {
-      r = { name: name, meta: meta };
+      r = { name, meta };
```
Agree, I have updated it and its name.

curq commented on src/plugins/console/public/lib/autocomplete/components/url_pattern_matcher.ts:
```diff
@@ -50,24 +61,26 @@ import { FullRequestComponent } from './full_request_component';
+      this.methodsData[method] = {
         rootComponent: new SharedComponent('ROOT'),
         parametrizedComponentFactories: parametrizedComponentFactories || {
           getComponent: () => {},
```
methodsData have the following type ```private methodsData: Record<string, MethodData>;```
```
interface MethodData {
  rootComponent: SharedComponent;
  parametrizedComponentFactories: ParametrizedComponentFactories | DefaultComponentFactories;
}

interface DefaultComponentFactories {
  getComponent: () => void;
}
```

curq commented:
@ananzh Thank you very much for your feedback! I have answered your comments and updated the code. Can you please take a look? I also removed lodash in some places. I will remove the rest in a separate PR after all TS migration PRs will be done.

ananzh commented on CHANGELOG.md:
```diff
@@ -213,7 +213,11 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [I18n] Fix Listr type errors and error handlers ([#3629](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3629))
 - [Multiple DataSource] Refactor dev tool console to use opensearch-js client to send requests ([#3544](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3544))
 - [Multiple DataSource] Present the authentication type choices in a drop-down ([#3693](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3693))
+- [Console] Remove unused ul element and its custom styling ([#3993](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3993))
```
Seems you include ur other PRs. CHANGELOG should only consider your current PR. For example, you should only include #4148. This might cause some future merge conflicts. For example if you merge #3798 first, then you would need to solve the conflict before merging this one. Including changes that are unrelated to the PR's can lead to confusion to others.

ananzh commented on src/plugins/console/public/lib/autocomplete/components/param_component.ts:
```diff
@@ -0,0 +1,51 @@
+/*
```
Think the license is still old.

ananzh commented on src/plugins/console/public/lib/autocomplete/types.ts:
```diff
@@ -0,0 +1,102 @@
+  addTemplate?: boolean;
+  indices?: string | string[];
+  types?: string | string[];
+  [key: string]: unknown;
```
make sense. thanks for the explain.

ananzh replied:
basically, context object can potentially have any number of different properties depending on the variable part of the URL...

ananzh commented on src/plugins/console/public/lib/autocomplete/components/url_pattern_matcher.ts:
```diff
@@ -50,24 +61,26 @@ import { FullRequestComponent } from './full_request_component';
+    // to avoid suggesting endpoints that are incompatible with the
+    // method that the user has entered.
+    this.methodsData = {};
     ['HEAD', 'GET', 'PUT', 'POST', 'DELETE'].forEach((method) => {
```
Yeah. Think this is nit.

ananzh reviewed (commented):
@curq thanks for fixing all the comments. This is definitely a lot of work. Could you update the CHANGELOG to only have #4148? Then update/rebase it and check all the tests. We do see cypress test and ftr tests failed recently, just verify if any tests failure is related to this PR. I could approve it then. 

This is great! You must spend quite a bit of time on it verifying all the types. We really appreciate your help~

joshuarrrr commented:
All checks pass.

joshuarrrr commented on CHANGELOG.md:
```diff
@@ -63,6 +63,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Vis Colors] [Maps] Replace hardcoded color to OUI color in `maps_legacy` plugin ([#4294](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4294))
 - [Vis Colors] [TSVB] Update default color in `vis_type_timeseries` to use `ouiPaletteColorBlind()[0]`([#4363](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4363))
 - [Vis Colors] [Timeline] Replace `vis_type_timeline` colors with `ouiPaletteColorBlind()` ([#4366](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4366))
+- [Console] Migrating `/lib/autocomplete/` Module to TypeScript ([#4148](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148))
```
```suggestion
- [Console] Migrate `/lib/autocomplete/` module to TypeScript ([#4148](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148))
```

joshuarrrr reviewed (commented):
@curq Just a few questions

curq commented on src/plugins/console/public/lib/autocomplete/components/constant_component.ts:
```diff
@@ -30,27 +30,31 @@
     this.options = options || [name];
   }
-  getTerms() {
+  getTerms(context?: AutoCompleteContext, editor?: CoreEditor | null): string[] | Term[] {
```
In both cases there are places where explicitly null is used.
For `constructor` https://github.com/opensearch-project/OpenSearch-Dashboards/blob/adb538ef73b7dd9d5488de6cb46eb7f8a7782661/src/plugins/console/public/lib/autocomplete/body_completer.js#L258
and here is the case when it is undefined. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/adb538ef73b7dd9d5488de6cb46eb7f8a7782661/src/plugins/console/public/lib/autocomplete/body_completer.js#L270

For the `getTerms` null case:
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/adb538ef73b7dd9d5488de6cb46eb7f8a7782661/src/plugins/console/public/lib/autocomplete/engine.js#L147
it is from populateContext method that can receive null context.
Specifically, in this test: 
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/adb538ef73b7dd9d5488de6cb46eb7f8a7782661/src/plugins/console/public/lib/autocomplete/__jest__/url_params.test.js#L61-L67
In case of undefined, there are several components that extend from `ConstantComponent` that do not pass arguments. for example: 
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/adb538ef73b7dd9d5488de6cb46eb7f8a7782661/src/plugins/console/public/lib/autocomplete/components/full_request_component.ts#L34-L43

So I believe the distinctions does matter, what do you think?

joshuarrrr replied:
Yeah, that seems fine for now.

curq commented on src/plugins/console/public/lib/autocomplete/body_completer.ts:
```diff
@@ -112,32 +136,45 @@ class ScopeResolver extends SharedComponent {
+      }
+    }
+    return [];
+  } else if (typeof description !== 'string' && description.__template) {
```
As I remember, I did this way to preserve completely the same behavior, because of this part
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/adb538ef73b7dd9d5488de6cb46eb7f8a7782661/src/plugins/console/public/lib/autocomplete/body_completer.js#L173-L179
I think we can remove the test and just leave string check because I don't think that `!/^\{.*\}$/.test(description)` does anything, because even if tests fails in the next else statement the description is returned anyway.
I refactored it to reflect that in my latest commit, you can recheck this part to see if everything is good to go.

curq commented:
@joshuarrrr Thank for the review, I responded to your comments, could you please check it?
